docs/error-handling.md

```diff
@@ -0,0 +1,190 @@
+# Error handling and logging
+
+## Logging errors
+
+The principle is that errors are logged when they are handled. If you
+just propagate an error to the caller in a function, you don't need to
+log it; the caller will. But if you "swallow" an error in a function,
+you *must* log it (if it needs to be logged at all).
+
+For example:
+
+```rust
+fn read_motd_file() -> std::io::Result<String> {
+    let mut f = File::open("/etc/motd")?;
+    let mut result = String::new();
+    f.read_to_string(&mut result)?;
+    result
+}
+```
+
+Opening or reading the file could fail, but there is no need to log
+the error here. The function merely propagates the error to the
+caller, and it is up to the caller to log the error or propagate it
+further, if the failure is not expected. But if, for example, it is
+normal that the "/etc/motd" file doesn't exist, the caller can choose
+to silently ignore the error, or log it in INFO or DEBUG level:
+
+```rust
+fn get_message_of_the_day() -> String {
+    // Get the motd from /etc/motd, or return the default proverb
+    match read_motd_file() {
+        Ok(motd) => motd,
+        Err(err)  => {
+            // It's normal that /etc/motd doesn't exist, but if we fail to
+            // read it for some other reason, that's unexpected. The message
+            // of the day isn't very important though, so we just WARN and
+            // continue with the default in any case.
+            if err.kind() != std::io::ErrorKind::NotFound {
+                 tracing::warn!("could not read \"/etc/motd\": {err:?}");
+            }
+            "An old error is always more popular than a new truth. - German proverb"
+        }
+    }
+}
+```
+
+## Error types
+
+We use the `anyhow` crate widely. It contains many convenient macros
```

Comment:
I have a different interpretation which involves two layers of thinking:

* On the surface, `.context(..)` [is](https://docs.rs/anyhow/latest/src/anyhow/context.rs.html#90-105) `ok_or_else(|| anyhow!(...))`. Which means that every time I see the latter, the question pops up: "why this again?". The name _might_ (but read below) be controversial, but the method itself is convenient.

* > it's unclear when reading Option::context that it's the source of the error instead of something providing context to an existing error

Oh, but there's an error and it's `None`! This is the source we've been looking for (although it's very shy and prefers to keep its secrets). The method does what it claims to do: it attaches a context to _an error_. Even the `eyre`'s doc explicitly says that: _anyhow avoids this issue by using a **different mental model** where you’re adding “context” to an error, though this not a mental model for error handling that eyre agrees with._

Clearly, you're accustomed to the `eyre`'s model, but this doesn't necessarily mean that the other model is wrong. Especially considering the first bullet.

Reply:
> Clearly, you're accustomed to the `eyre`'s model, but this doesn't necessarily mean that the other model is wrong.

Fair enough :)

> Oh, but there's an error and it's `None`!

I think this is the part that I disagree with, but if I'm in the minority here that's ok